src/adaptors/adaptor_tls.c

```diff
@@ -723,7 +729,289 @@ bool qd_tls_has_output(const qd_tls_t *tls)
+                 qd_tls_take_output_data_cb_t *take_output_cb,
+                 void                         *take_output_context,
+                 qd_adaptor_buffer_list_t     *input_data,
+                 uint64_t                     *input_data_count)
+{
+    bool work;
+    *input_data_count = 0;
+
+    assert(tls);
+    assert(raw_conn);
+
+    do {
+        size_t          capacity;
+        size_t          taken;
+        size_t          given;
+        pn_raw_buffer_t pn_buf_desc;
+        uint64_t        total_octets;
+
+        // Loop until no more work can be done. "work" is considered true whenever the TLS layer produces output or
+        // opens up capacity for more input
+
+        work = false;
+
+        // give empty buffers for holding encrypted output from TLS
+
+        capacity = pn_tls_get_encrypt_output_buffer_capacity(tls->tls_session);
+        while (capacity-- > 0) {
+            (void) qd_adaptor_buffer_raw(&pn_buf_desc);
+            given = pn_tls_give_encrypt_output_buffers(tls->tls_session, &pn_buf_desc, 1);
+            (void) given;
+            assert(given == 1);
+        }
+
+        // push any unencrypted output data from the protocol adaptor into TLS for encryption.
+
+        if (pn_tls_is_secure(tls->tls_session)) {
+            assert(take_output_cb);
+            capacity = pn_tls_get_encrypt_input_buffer_capacity(tls->tls_session);
+
+            if (capacity > 0) {
+                qd_adaptor_buffer_list_t ubufs = DEQ_EMPTY;
+
+                int64_t out_octets = take_output_cb(take_output_context, &ubufs, capacity);
+                if (out_octets > 0) {
+                    assert(!DEQ_IS_EMPTY(ubufs) && DEQ_SIZE(ubufs) <= capacity);
+                    qd_log(tls->log_source,
+                           QD_LOG_TRACE,
+                           "[C%" PRIu64 "] %" PRIi64 " unencrypted octets taken by TLS for encryption (%zu buffers)",
+                           tls->conn_id,
+                           out_octets,
+                           DEQ_SIZE(ubufs));
+                    qd_adaptor_buffer_t *abuf = DEQ_HEAD(ubufs);
+                    while (abuf) {
+                        DEQ_REMOVE_HEAD(ubufs);
+                        qd_adaptor_buffer_pn_raw_buffer(&pn_buf_desc, abuf);
+                        given = pn_tls_give_encrypt_input_buffers(tls->tls_session, &pn_buf_desc, 1);
+                        assert(given == 1);
+                        abuf = DEQ_HEAD(ubufs);
+                    }
+                } else {
+                    assert(DEQ_IS_EMPTY(ubufs));
+                    if (out_octets == QD_TLS_EOS && tls->output_eos == false) {
+                        pn_tls_close_output(tls->tls_session);
+                        tls->output_eos = true;
+                        qd_log(tls->log_source, QD_LOG_TRACE, "[C%" PRIu64 "] EOS signalled: closing TLS output",
+                               tls->conn_id);
+                    }
+                }
+            }
+        }
```

Comment:
The doc for pn_tls_is_secure sugests that it will also return false on some kind of error.  Just checking that that is acounted for below where you inspect tls_error.

Reply:
@ssorj see the latest commit on this PR. Added some comments to explain the "not secure" handling.  Lemme know what you think.